### PR TITLE
Typed Column Accessor

### PR DESCRIPTION
**Pull request by adi64**
Alright, second try (after #985) :wink:

**What's the problem?**
A ReferenceColumn does not know the concrete type of the columns it references to. Currently, the ReferenceColumnIterable just uses the subscript-operator to fetch AllTypeVariants. An alternative would be to do a `resolve_column_type` on every access -- but we can't cache the resolved column types because the only common base class of columns is BaseColumn. I tried this and it was prohibitively expensive.

**Alright, what does this do?**
This introduces a typed column accessor as proposed by @mjendruk. All concrete column types[1] know their data type (`T`) and can therefore implement a method to retrieve a value directly. In contrast to #985, here no multiple inheritance is used, but an implicit interface that has to be implemented by all concrete column types:
`const std::optional<T> get_typed_value(ChunkOffset chunk_offset) const`

[1] with the notable exception of a ReferenceColumn, so here we fall back to the subscript operator and AllTypeVariant.

**Show me the numbers!**
On sidon, scale factor 1.0, chunk size 1M. TPC-H 20 was omitted because it takes forever.

```
+-----------+---------------------+------+---------------------+------+--------+--------------------------------+
| Benchmark | prev. iter/s        | runs | new iter/s          | runs | change | p-value (significant if <0.01) |
+-----------+---------------------+------+---------------------+------+--------+--------------------------------+
| TPC-H 1   | 0.182072848081589   | 10   | 0.196682870388031   | 10   | +8%    |    (run time too short) 0.0000 |
| TPC-H 2   | 1.28981590270996    | 10   | 1.20346105098724    | 10   | -7%    |    (run time too short) 0.0203 |
| TPC-H 3   | 1.82134211063385    | 10   | 2.08979988098145    | 10   | +15%   |    (run time too short) 0.0000 |
| TPC-H 4   | 0.0190980490297079  | 10   | 0.0190930049866438  | 10   | -0%    |                         0.9427 |
| TPC-H 5   | 1.2901828289032     | 10   | 1.3925176858902     | 10   | +8%    |    (run time too short) 0.0000 |
| TPC-H 6   | 3.14697170257568    | 10   | 3.64119577407837    | 10   | +16%   |    (run time too short) 0.0006 |
| TPC-H 7   | 0.262979984283447   | 10   | 0.316446781158447   | 10   | +20%   |    (run time too short) 0.0000 |
| TPC-H 8   | 0.027232164517045   | 10   | 0.0408398732542992  | 10   | +50%   |                         0.0000 |
| TPC-H 9   | 0.127553656697273   | 10   | 0.147799000144005   | 10   | +16%   |                         0.0000 |
| TPC-H 10  | 1.17688453197479    | 10   | 1.35261869430542    | 10   | +15%   |    (run time too short) 0.0000 |
| TPC-H 11  | 3.58690237998962    | 10   | 3.76866149902344    | 10   | +5%    |    (run time too short) 0.0000 |
| TPC-H 12  | 1.92889964580536    | 10   | 2.19707846641541    | 10   | +14%   |    (run time too short) 0.0000 |
| TPC-H 13  | 1.22080421447754    | 10   | 1.28417778015137    | 10   | +5%    |    (run time too short) 0.0000 |
| TPC-H 14  | 3.79250001907349    | 10   | 4.52005767822266    | 10   | +19%   |    (run time too short) 0.0021 |
| TPC-H 15  | 1.35429155826569    | 10   | 1.55974185466766    | 10   | +15%   |    (run time too short) 0.0000 |
| TPC-H 16  | 0.581602931022644   | 10   | 0.650469899177551   | 10   | +12%   |    (run time too short) 0.0001 |
| TPC-H 17  | 0.030084079131484   | 10   | 0.0298776775598526  | 10   | -1%    |                         0.2291 |
| TPC-H 18  | 0.0025155465118587  | 10   | 0.00253599369898438 | 5    | +1%    |       (not enough runs) 0.1212 |
| TPC-H 19  | 0.104484625160694   | 10   | 0.115418806672096   | 10   | +10%   |                         0.0000 |
| TPC-H 21  | 0.0130579061806202  | 10   | 0.0125842411071062  | 10   | -4%    |                         0.0000 |
| TPC-H 22  | 0.00765585154294968 | 10   | 0.0077105900272727  | 10   | +1%    |                         0.4148 |
| average   |                     |      |                     |      | +10%   |                                |
+-----------+---------------------+------+---------------------+------+--------+--------------------------------+
```
(Apparently, we are now *too fast* - If needed, I can re-run with more iterations)

This is interestingly *slightly* slower than the multiple inheritance version:
```
+-----------+---------------------+------+--------------------+------+--------+--------------------------------+ 
| Benchmark | prev. iter/s        | runs | new iter/s         | runs | change | p-value (significant if <0.01) | 
+-----------+---------------------+------+--------------------+------+--------+--------------------------------+ 
| TPC-H 1   | 0.182072848081589   | 10   | 0.193810909986496  | 10   | +6%    |    (run time too short) 0.0000 | 
| TPC-H 2   | 1.28981590270996    | 10   | 1.36888694763184   | 10   | +6%    |    (run time too short) 0.0000 | 
| TPC-H 3   | 1.82134211063385    | 10   | 2.10876870155334   | 10   | +16%   |    (run time too short) 0.0000 | 
| TPC-H 4   | 0.0190980490297079  | 10   | 0.0189696215093136 | 10   | -1%    |                         0.1241 | 
| TPC-H 5   | 1.2901828289032     | 10   | 1.42408454418182   | 10   | +10%   |    (run time too short) 0.0000 | 
| TPC-H 6   | 3.14697170257568    | 10   | 3.61715841293335   | 10   | +15%   |    (run time too short) 0.0028 | 
| TPC-H 7   | 0.262979984283447   | 10   | 0.326312243938446  | 10   | +24%   |    (run time too short) 0.0000 | 
| TPC-H 8   | 0.027232164517045   | 10   | 0.0431611053645611 | 10   | +58%   |                         0.0000 | 
| TPC-H 9   | 0.127553656697273   | 10   | 0.152487844228745  | 10   | +20%   |                         0.0000 |             
| TPC-H 10  | 1.17688453197479    | 10   | 1.34591639041901   | 10   | +14%   |    (run time too short) 0.0000 | 
| TPC-H 11  | 3.58690237998962    | 10   | 3.87801384925842   | 10   | +8%    |    (run time too short) 0.0000 | 
| TPC-H 12  | 1.92889964580536    | 10   | 2.23412299156189   | 10   | +16%   |    (run time too short) 0.0000 | 
| TPC-H 13  | 1.22080421447754    | 10   | 1.30718040466309   | 10   | +7%    |    (run time too short) 0.0000 | 
| TPC-H 14  | 3.79250001907349    | 10   | 4.58873081207275   | 10   | +21%   |    (run time too short) 0.0007 | 
| TPC-H 15  | 1.35429155826569    | 10   | 1.52278256416321   | 10   | +12%   |    (run time too short) 0.0000 | 
| TPC-H 16  | 0.581602931022644   | 10   | 0.640903234481812  | 10   | +10%   |    (run time too short) 0.0003 | 
| TPC-H 17  | 0.030084079131484   | 10   | 0.0300237312912941 | 10   | -0%    |                         0.7074 |
| TPC-H 18  | 0.0025155465118587  | 10   | 0.0025500375777483 | 5    | +1%    |       (not enough runs) 0.0317 |             
| TPC-H 19  | 0.104484625160694   | 10   | 0.116768665611744  | 10   | +12%   |                         0.0000 |                                                                                       
| TPC-H 21  | 0.0130579061806202  | 10   | 0.0131742153316736 | 10   | +1%    |                         0.0259 |
| TPC-H 22  | 0.00765585154294968 | 10   | 0.0079096956178546 | 10   | +3%    |                         0.0007 |
| average   |                     |      |                    |      | +12%   |                                | 
+-----------+---------------------+------+--------------------+------+--------+--------------------------------+ 
```

### CODE REVIEW

File: src/lib/operators/sort.cpp [hunk 1]
```diff
@@ -88,15 +90,34 @@ class Sort::SortImplMaterializeOutput {
...
           const auto [chunk_id, chunk_offset] = _row_id_value_vector->at(row_index).first;  // NOLINT
 
-          const auto column = _table_in->get_chunk(chunk_id)->get_column(column_id);
+          auto& column_ptr_and_typed_ptr_pair = column_ptr_and_accessor_by_chunk_id[chunk_id];
+          auto& base_column = column_ptr_and_typed_ptr_pair.first;
+          auto& accessor = column_ptr_and_typed_ptr_pair.second;
+
+          if (!base_column) {
+            base_column = _table_in->get_chunk(chunk_id)->get_column(column_id);
+            accessor = std::move(get_column_accessor<ColumnDataType>(base_column));
```
Comment (mrks) [inline src/lib/operators/sort.cpp]: why do you move a return value?

File: src/lib/storage/run_length_column.cpp [hunk 2]
```diff
@@ -36,15 +36,24 @@ std::shared_ptr<const pmr_vector<ChunkOffset>> RunLengthColumn<T>::end_positions
 template <typename T>
 const AllTypeVariant RunLengthColumn<T>::operator[](const ChunkOffset chunk_offset) const {
   PerformanceWarning("operator[] used");
+  const auto typed_value = get_typed_value(chunk_offset);
+  if (!typed_value.has_value()) {
+    return NULL_VALUE;
+  }
+  return *typed_value;
+}
```
Comment (mrks) [inline src/lib/storage/run_length_column.cpp]: The doc for `operator[]` should have a comment telling people to use the new version whenever (?) they know the type.

File: src/lib/operators/sort.cpp (hunk 1 above)
Comment (adi64) [reply src/lib/operators/sort.cpp]: good point, fixed.

File: src/lib/storage/run_length_column.cpp (hunk 2 above)
Comment (adi64) [reply src/lib/storage/run_length_column.cpp]: I added this to base_column.hpp, is that okay?
```
  // Returns the value at a given position. This is considerably slow since it returns an AllTypeVariant.
  // If you know the column type and its stored data type, consider using get_typed_value(chunk_offset) instead.
  virtual const AllTypeVariant operator[](const ChunkOffset chunk_offset) const = 0;
```

File: src/lib/operators/sort.cpp [hunk 3]
```diff
@@ -88,15 +90,34 @@ class Sort::SortImplMaterializeOutput {
         auto column_it = columns_out.begin();
         auto chunk_it = output_columns_by_chunk.begin();
         auto chunk_offset_out = 0u;
+
+        auto column_ptr_and_accessor_by_chunk_id =
```
Comment (marcelja) [inline src/lib/operators/sort.cpp]: Maybe it would be better if "sort" would not materialize the output. That's not part of this PR though.

**Comment by mrks** [comment]
@mrzzzrm - you said that you would prefer this one but still had some small things to add?

File: src/lib/storage/column_accessor.hpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+#include "types.hpp"
+
+namespace opossum {
+
+/**
+ * This is the base class for all ColumnAccessor types.
+ * It provides the common interface to access individual values of a column.
+ */
+template <typename T>
+class BaseColumnAccessor {
```
Comment (mrzzzrm) [inline src/lib/storage/column_accessor.hpp]: Even though it's minimal, this violates the one-class-per-compilation-unit rule
please transfer this to its own header.

File: src/lib/storage/column_accessor.hpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+ */
+template <typename T, typename ColumnType>
+class ColumnAccessor : public BaseColumnAccessor<T> {
+ public:
+  explicit ColumnAccessor(const ColumnType& column) : _column{column} {}
+
+  const std::optional<T> access(ChunkOffset offset) const final { return _column.get_typed_value(offset); }
+
+ protected:
+  const ColumnType& _column;
```
Comment (mrzzzrm) [inline src/lib/storage/column_accessor.hpp]: Any reason why this should hold a smart pointer?

File: src/lib/storage/column_accessor.hpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+/**
+ * Partial template specialization for ReferenceColumns.
+ * Since ReferenceColumns don't know their 'T', this uses the subscript operator as a fallback.
+ */
+template <typename T>
+class ColumnAccessor<T, ReferenceColumn> : public BaseColumnAccessor<T> {
+ public:
+  explicit ColumnAccessor(const ReferenceColumn& column) : _column{column} {}
+  const std::optional<T> access(ChunkOffset offset) const final {
+    const auto all_type_variant = _column[offset];
```
Comment (mrzzzrm) [inline src/lib/storage/column_accessor.hpp]: A ColumnAccessor should never be created on a ReferenceColumn, am I right? Just `Fail()` here, if I am.

File: src/lib/storage/column_accessor.hpp [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+ protected:
+  const ReferenceColumn& _column;
+};
+
+/**
+ * Utility method to create a ColumnAccessor for a given BaseColumn.
+ */
+template <typename T>
+std::unique_ptr<BaseColumnAccessor<T>> get_column_accessor(const std::shared_ptr<const BaseColumn>& column) {
```
Comment (mrzzzrm) [inline src/lib/storage/column_accessor.hpp]: `create_column_accessor`

File: src/lib/storage/value_column.cpp [hunk 8]
```diff
@@ -84,11 +92,26 @@ void ValueColumn<T>::append(const AllTypeVariant& val) {
...
   }
 
-  Assert(!is_null, "ValueColumns is not nullable but value passed is null.");
+  Assert(!is_null, "ValueColumn is not nullable but value passed is null.");
 
   _values.push_back(type_cast<T>(val));
 }
 
+template <typename T>
+void ValueColumn<T>::append_typed_value(const std::optional<T> value_or_null) {
```
Comment (mrzzzrm) [inline src/lib/storage/value_column.cpp]: Do you really need this, or can you just pre-build the null/value vectors in Sort? I am really against needlessly extending the interface of our core storage classes.

File: src/test/storage/column_accessor_test.cpp [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    tbl->append_chunk(chunk);
+
+    pos_list = std::make_shared<PosList>(PosList{
+        {RowID{ChunkID{0}, ChunkOffset{1}}, RowID{ChunkID{0}, ChunkOffset{2}}, RowID{ChunkID{0}, ChunkOffset{0}}}});
+
+    rc_int = std::make_shared<ReferenceColumn>(tbl, ColumnID{0}, pos_list);
+    rc_str = std::make_shared<ReferenceColumn>(tbl, ColumnID{1}, pos_list);
+  }
+
+  std::shared_ptr<BaseValueColumn> vc_int = nullptr;
```
Comment (mrzzzrm) [inline src/test/storage/column_accessor_test.cpp]: minor: `= nullptr` is not needed.

File: src/lib/storage/column_accessor.hpp (hunk 4 above)
Comment (adi64) [reply src/lib/storage/column_accessor.hpp]: added `base_column_accessor.hpp`

File: src/lib/storage/column_accessor.hpp (hunk 5 above)
Comment (adi64) [reply src/lib/storage/column_accessor.hpp]: What do you mean? It does not hold a smart pointer.

If you mean why it does *not* hold a smart pointer, two reasons:
1. I don't see a reason why it should, as it would add indirection without any added use
2. `resolve_column_type` gives a const-ref to the concrete column type and `create_column_accessor` (which is a thin wrapper around that) is the usual way to construct a ColumnAccessor.

File: src/lib/storage/column_accessor.hpp (hunk 6 above)
Comment (adi64) [reply src/lib/storage/column_accessor.hpp]: The motivating use case for ColumnAccessors is the ReferenceColumnIterable and hopefully that will not try to access other ReferenceColumns, correct. However, I don't see why this should fail if it works just fine and isn't slower than the usual `operator[]`.

File: src/lib/storage/column_accessor.hpp (hunk 7 above)
Comment (adi64) [reply src/lib/storage/column_accessor.hpp]: done

File: src/test/storage/column_accessor_test.cpp (hunk 9 above)
Comment (adi64) [reply src/test/storage/column_accessor_test.cpp]: done

File: src/lib/storage/column_accessor.hpp (hunk 6 above)
Comment (mrks) [reply src/lib/storage/column_accessor.hpp]: Maybe PerformanceWarning as middle ground?

File: src/lib/storage/column_accessor.hpp (hunk 5 above)
Comment (mrzzzrm) [reply src/lib/storage/column_accessor.hpp]: I meant "not hold", of course. 

We use smart pointers over references for class members everywhere else, so I would have done it here as well for consistency reasons. since `resolve_column_type` doesn't give you a smart pointer to the column, I guess I have to let this one pass

File: src/lib/storage/value_column.cpp (hunk 8 above)
Comment (mrzzzrm) [reply src/lib/storage/value_column.cpp]: Did you push? Sort still seems to use `append_typed_value` as far as I can tell.

File: src/lib/storage/column_accessor.hpp (hunk 6 above)
Comment (adi64) [reply src/lib/storage/column_accessor.hpp]: done

File: src/lib/storage/value_column.cpp (hunk 8 above)
Comment (adi64) [reply src/lib/storage/value_column.cpp]: Yes, see [here](https://github.com/hyrise/hyrise/commit/e5f8a8e9a5c045fe640460e37ce762c90227c939#diff-a735884cb8be5955a5069458e2908620L112)
